### ISSUE TITLE
[XL] Reservations - Move SAM to CDK

### ISSUE THREAD

**Issue by Dianadec** [bcgov/reserve-rec-api#42]
#### Description:

NOTE: This ticket will only be for migrating SAM to CDK for Reservations stuff. Other apps will stay with SAM. go sam!

This ticket is the next step that came out of 37: https://github.com/bcgov/reserve-rec-api/issues/37 

 **Additional Context:**

The first time AWS infrastructure was used by this team was for DUP back in 2021. Our infrastructure was deployed by a Terraform script that was written by our then-devops developer in that really small 5 week window. The script worked, but was hardly scalable. A lot of effort had to be put in to make any changes to our infrastructure.

Restrictions imposed by the AWS team prevented us from moving away from Terraform by the time our next project, A&R came around. After A&R was launched there were changes made that allowed us to move away from Terraform, which was hampering development greatly by that point, so we moved to the quickest alternative, which was SAM. We'd spent a bit of time investigating the feasibility of SAM because we were so fed up with Terraform so we jumped at the opportunity to use it with the Data Register. SAM's been working for us great ever since. So great that we backported it to DUP and A&R so we could make changes to infrastructure quickly on those projects.

So that's a brief history of why we are on SAM for our projects right now and didn't jump directly to CDK.

Notes from Dec 4 Retro:
Build times locally suck. - Might be a way to fix it but difficult.... lambda having own containers. Perhaps moving to CDK will solve this problem?? could look into having sep stack for everyone in dev??? then would need to deploy every time you make a change. Deploying every time you make a change might be faster than a rebuild...
Soo... **See how building locally works after cdk.**

#### Development Checklist:

CDK templates can be easily generated from existing SAM templates OR CloudFormation Stacks: https://docs.aws.amazon.com/cdk/v2/guide/migrate.html#:~:text=Migrate%20from%20an%20AWS%20SAM,migrate%20from%20the%20deployed%20stack.

Install AWS CDK package npm install -g aws-cdk
Log into Nimbus, paste credentials into terminal
cdk migrate --from-stack --stack-name "reserve-rec"

**December 3rd notes from #37;** 

I've done enough exploring with AWS CDK the past few days to be convinced that a migration is easily possible, albeit a little time consuming. Additionally, we would have to spin down the existing SAM deployment simultaneously, and since that is currently supporting some front end work, I haven't gone through with this change yet.

Currently our reserve-rec deployment using SAM is serving us fine but there will come a point where our rate of growth will be impeded by the monolithic structure of SAM templates - and by that time, the effort required to migrate to CDK will have grown exponentially. I recommend we set aside some time soon to do this migration, as it will only get more difficult the more we progress. I consider this issue tech debt - we should do this, it's just a matter of when. In the upcoming weeks there is a no-meeting week and the Christmas holidays which would perhaps be opportune times for completing this work.

#### Dependencies

- Blocked by
- Blocking

**Relevant documentation as reference**

**Definition of Ready**

- [ ] Acceptance criteria are included
- [ ] Wireframes are included (if applicable)
- [ ] Design / Solution is accepted by Product Owner (if applicable)
- [ ] Dependencies are identified (technical, business, regulatory/policy)
- [ ] Story has been estimated (under 13 pts)

**Definition of Done** 
- In progress:
  - [ ] Acceptance criteria are tested (Functionality meets the acceptance criteria defined in the ticket)
  - [ ] UI meets accessibility requirements
  - [ ] Unit tests are written
  - [x] Work is traceable in GitHub
  - [x] PR linked to ticket number
  - [ ] If needed/required - Dev adds flag/label to highlight any migration steps necessary prior to PROD deployment
- Code review:
  - [x] Code is peer reviewed and has passed CI/CD tests
- QA:
  - [ ] Acceptance criteria are tested (Functionality meets the acceptance criteria defined in the ticket)
  - [ ] Code is potentially shippable to the production environment
  - [ ] Functional features have been tested and passed by QA
  - [ ] UI components tested by designer
  - [ ] Code is deployed to PROD when moved to 'done' column (unless requested otherwise by PO)
- PO Review:
  - [ ] Acceptance criteria are tested (Functionality meets the acceptance criteria defined in the ticket)
  - [x] Reviewed and approved by Product Owner

#### Notes:
- 
-

**Comment by LindsayMacfarlane** [comment]
@cameronpettit - are there any additional details you can add to this ticket?

**Comment by Dianadec** [comment]
@LindsayMacfarlane  and @cameronpettit I transferred notes and what I think might be dev checklist-y things from #37 into this ticket. Please review and add anything else as needed. I'll put this up for estimates as well.

**Comment by cameronpettit** [comment]
Seems complete to me!

**Comment by Dianadec** [comment]
in progress - currently working on gh actions

**Comment by Dianadec** [comment]
Once cognito is done, might move to the area of work to unblock pub/sub ticket.

**Comment by Dianadec** [comment]
@LindsayMacfarlane fyi just chatting with @danieltruong  - the Front-end CDK things still need to be done. This may need to be captured in another ticket(s) but they also may complete it sooner. We will check in at standup tomorrow when going through the board of tickets.

**Comment by danieltruong** [comment]
As of 2025-02-12, CDK for reserve-rec-api is done. Any tickets related to API should be unblocked. We will continue working on converting reserve-rec-public and reserve-rec-admin to use CDK.

**Comment by Dianadec** [comment]
Note about dividing FE: https://teams.microsoft.com/l/message/19:163facd69a2e4fe1a14eac3ae95fb3c4@thread.tacv2/1740090056367?tenantId=6fdb5200-3d0d-4a8a-b036-d3685e359adc&groupId=aadf267c-d6ab-4c8c-9cff-e574713575e6&parentMessageId=1740090056367&teamName=External%3A%20Team%20Osprey&channelName=Dev%20Chat&createdTime=1740090056367

**Comment by cameronpettit** [comment]
Reserve-rec-public is now done, and the Wiki has been updated with some information on how to setup a local environment using CDK. Reserve-rec-admin is also complete but we have yet to overwrite the main branch with it

**Comment by Dianadec** [comment]
This is now the way we do CI/CD. over to you Lindsay!

**Comment by LindsayMacfarlane** [comment]
Woohoo!!!! Well done @cameronpettit and @danieltruong - this was such a huge piece of work and a huge thank you and congratulations is in order....so CONGRATULATIONS!
